google/cloud/bigtable/mutations_batcher.py

```diff
@@ -46,61 +177,318 @@ class MutationsBatcher:
 
-    async def flush(self):
+    def _add_exceptions(self, excs: list[Exception]):
         """
-        Send queue over network in as few calls as possible
+        Add new list of exceptions to internal store. To avoid unbounded memory,
+        the batcher will store the first and last _exception_list_limit exceptions,
+        and discard any in between.
+        """
+        self._exceptions_since_last_raise += len(excs)
+        if excs and len(self._oldest_exceptions) < self._exception_list_limit:
+            # populate oldest_exceptions with found_exceptions
+            addition_count = self._exception_list_limit - len(self._oldest_exceptions)
+            self._oldest_exceptions.extend(excs[:addition_count])
+            excs = excs[addition_count:]
+        if excs:
+            # populate newest_exceptions with remaining found_exceptions
+            keep_count = self._exception_list_limit - len(excs)
+            self._newest_exceptions = (
+                excs[-self._exception_list_limit :]
+                + self._newest_exceptions[:keep_count]
+            )
+
+    def _raise_exceptions(self):
+        """
+        Raise any unreported exceptions from background flush operations
 
         Raises:
-        - MutationsExceptionGroup if any mutation in the batch fails
+          - MutationsExceptionGroup with all unreported exceptions
         """
-        raise NotImplementedError
+        if self._oldest_exceptions or self._newest_exceptions:
+            oldest, self._oldest_exceptions = self._oldest_exceptions, []
+            newest, self._newest_exceptions = self._newest_exceptions, []
+            entry_count, self._entries_processed_since_last_raise = (
+                self._entries_processed_since_last_raise,
+                0,
+            )
+            exc_count, self._exceptions_since_last_raise = (
+                self._exceptions_since_last_raise,
+                0,
+            )
+            raise MutationsExceptionGroup.from_truncated_lists(
+                first_list=oldest,
+                last_list=newest,
+                total_excs=exc_count,
+                entry_count=entry_count,
+            )
 
     async def __aenter__(self):
         """For context manager API"""
-        raise NotImplementedError
+        return self
 
     async def __aexit__(self, exc_type, exc, tb):
         """For context manager API"""
-        raise NotImplementedError
+        await self.close()
 
     async def close(self):
         """
         Flush queue and clean up resources
         """
-        raise NotImplementedError
+        self.closed = True
+        self._flush_timer.cancel()
+        self._schedule_flush()
+        await asyncio.gather(*self._flush_jobs, return_exceptions=True)
+        try:
+            await self._flush_timer
+        except asyncio.CancelledError:
+            pass
+        # raise unreported exceptions
+        self._raise_exceptions()
+        atexit.unregister(self._on_exit)
```

Comment:
I think you might want to do this before raising exceptions

Reply:
good catch